Build and link to tcmalloc statically

brson opened the pull request:
This makes gperftools a submodule, adds a tcmalloc-sys crate, and builds and links to tcmalloc.

If somebody really wants to use a local tcmalloc they can enable the `no-build` feature.

Note that I've only tested on x86_64-linux-unknown-gnu, and so I whitelisted that target in the build script &mdash; for any other target the script will fail.

This also bumps the version number. If you accept this PR I hope you will publish a new revision.

brson committed:
Build and link tcmalloc statically

This adds a tcmalloc-sys crate. If the "bundled" feature
is enabled it builds tcmalloc.

jmcomets commented:
Thanks for the PR! I do have a few minor review notes on the `build.rs` script.

Aside from the code itself, I'm wondering if a submodule is the right way to go here. For example, [tensorflow-sys](https://github.com/tensorflow/rust/blob/master/tensorflow-sys/build.rs#L204) clones the repository and checks out the specific tag before building from there. If you have any extra info on that, I'd be interested, but I'll accept it anyways.

jmcomets commented on Cargo.toml:
```diff
@@ -13,4 +13,9 @@ documentation = "https://docs.rs/tcmalloc"
 
+[features]
+default = []
+no-build = ["tcmalloc-sys/no-build"]
```
I'd flip this option, building from source only when a feature is given (`bundled` sounds like a good name).

brson replied:
Will do.

jmcomets commented on tcmalloc-sys/build.rs:
```diff
@@ -0,0 +1,93 @@
+    info!("GPERFTOOLS_DIR={:?}", gperftools_dir);
+
+    if !TESTED.contains(&target.as_ref()) {
+        panic!("tcmalloc-rs has not been verified to work on target {}", target);
```
I don't think it's a good idea to `panic!` when we're building for an untested target. A warning should suffice.

brson replied:
Will do.

jmcomets commented on tcmalloc-sys/build.rs:
```diff
@@ -0,0 +1,93 @@
+    copy_options.overwrite = true;
+    copy_options.copy_inside = true;
+    fs_extra::dir::copy(Path::new("gperftools"), &gperftools_dir, &copy_options)
+        .expect("failed to copy jemalloc source code to OUT_DIR");
```
Is there a particular reason why this copy is required? (Can't you build directly in gperftools' source?)

jmcomets replied:
Side note: the message should mention `tcmalloc`, not `jemalloc` :grinning:

brson replied:
It's to avoid polluting the package source directory, which is, I think, a global resource.

Good catch on the message.

jmcomets commented on tcmalloc-sys/build.rs:
```diff
@@ -0,0 +1,93 @@
+}
+
+fn run(cmd: &mut Command) {
+    println!("running: {:?}", cmd);
```
For consistency, this should probably use the `info!` macro you defined above.

brson replied:
Yes.

brson commented:
@jmcomets Re submodule vs cloning during build, I am mostly copying from jemallocator here. The main reason to use a submodule and not do a clone that I can think of is that `git clone` only works for branches and tags, not arbitrary commits. But gperftools is well tagged so that doesn't really apply. I will change it to do a clone.

brson commented:
Thanks for the review.

I pushed a rebase that addresses all comments.

I deleted `info!` in favor of `println!`. "bundled" is now not the default, and the tcmalloc-sys crate isn't used when not "bundled".

I also removed the version bump. Since this is adding an optional feature I don't think a major bump is necessary. You can handle the versions.

jmcomets commented:
This looks great and works out of the box, well done. :+1:

jmcomets commented:
FYI there may be some notification spam here while I get this published to crates.io, don't mind it too much
